File: app/src/main/java/com/joeshuff/dddungeongenerator/generator/dungeon/Room.java
package com.joeshuff.dddungeongenerator.generator.dungeon;

import android.graphics.Point;
import com.joeshuff.dddungeongenerator.Logs;
import com.joeshuff.dddungeongenerator.generator.features.*;
import com.joeshuff.dddungeongenerator.generator.floors.DungeonSection;
import com.joeshuff.dddungeongenerator.generator.models.Rectangle;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Random;

public class Room  {

    static transient List<String> featureIntros = Arrays.asList("You've heard rumors that ",
            "You've discovered that ",
            "Maps have told you that ",
            "It became apparent that ",
            "You know that ");

    private transient DungeonSection thisSection;
    private int sectionStartX = 0;
    private int sectionStartY = 0;

    private static transient final int MIN_MEAN_ROOM_SIZE = 45;
    private static transient final int MAX_MEAN_ROOM_SIZE = 85;

    private static transient final int MIN_STD_DEV = 15;
    private static transient final int MAX_STD_DEV = 45;

    private static transient int MEAN_ROOM_SIZE = 65;
    private static transient int STD_DEV = 30;

    private int id;

    private String detail;

    private int startX;
    private int startY;

    private boolean isSelected = false;
    private boolean isRejected = false;

    private int width;
    private int height;

    static transient double DEFAULT_MONSTER_CHANCE = 0.15d;
    static transient double DEFAULT_TRAP_CHANCE = 0.2d;
    static transient double DEFAULT_TREASURE_CHANCE = 0.25d;
    static transient double DEFAULT_STAIRS_CHANCE = 0.05d;

    private int timesShifted = 0;

    transient Random rnd;

    List<RoomFeature> featureList = new ArrayList<>();

    transient List<Movement> movementHistory = new ArrayList<>();

    public static void changeRoomSize(float percentage) {
        MEAN_ROOM_SIZE = (int) (MIN_MEAN_ROOM_SIZE + ((MAX_MEAN_ROOM_SIZE - MIN_MEAN_ROOM_SIZE) * percentage));
        STD_DEV = (int) (MIN_STD_DEV + ((MAX_STD_DEV - MIN_STD_DEV) * percentage));
    }

    public void setId(int id) {
        this.id = id;
    }

    public Room(DungeonSection section, int id, Random rnd, Modifier modifier) {
        this.rnd = rnd;
        this.thisSection = section;
        this.sectionStartX = section.getStartPoint().x;
        this.sectionStartY = section.getStartPoint().y;

        this.id = id;

        Point center = randomPointInCircle((section.getWidth() / 4));

        this.height = (int) ((rnd.nextGaussian() * STD_DEV) + MEAN_ROOM_SIZE);
        this.width = (int) ((rnd.nextGaussian() * STD_DEV) + MEAN_ROOM_SIZE);

        this.detail = "This room is " + height + "ft x " + width + "ft\n" + RoomDetail.getDetail(rnd);

        startX = (section.getWidth() / 2) + center.x - (this.width / 2);
        startY = (section.getHeight() / 2) + center.y - (this.height / 2);

        generateFeatures(modifier);
    }

    public String getDetail() {
        return detail;
    }

    public String getFullRoomDescription() {
        List<String> descParts = getRoomDescription();

        String res = "";

        for (String d : descParts) {
            res = res + d + "\n";
        }

        return res;
    }

    public List<String> getRoomDescription() {
        List<String> description = new ArrayList<>();

        for (RoomFeature feature : getFeatureList()) {
            description.add(featureIntros.get(new Random().nextInt(featureIntros.size())) + feature.getFeatureDescription());
        }

        return description;
    }

    public void connectRoomTo(Room room, String connection, StairsFeature.DIRECTION direction) {
        StairsFeature newStairs = new StairsFeature(direction.opposite(), connection, room.getId(), room.getThisSection().getFloor().getLevel());
        featureList.add(newStairs);
    }

    private Point randomPointInCircle(int radius) {
        double t = 2.0f * Math.PI * rnd.nextDouble();
        double u = rnd.nextDouble() + rnd.nextDouble();
        double r;

        if (u > 1) {
            r = 2 - u;
        } else {
            r = u;
        }

        Point p = new Point();
        p.x = (int) (radius * r * Math.cos(t));
        p.y = (int) (radius * r * Math.sin(t));

        return p;
    }

    public boolean moveAwayFrom(Room room) {
        Point center = myCenter();
        Point otherCenter = room.myCenter();

        if (timesShifted >= 1000) {
            isRejected = true;
            return true;
        }

        Point diff = new Point(center.x - otherCenter.x, center.y - otherCenter.y);

        Rectangle intersection = room.meWithBorder().intersection(meWithBorder());

        if (intersection.isEmpty()) return false;
        if (diff.x == 0 && diff.y == 0) return false;

        int xOverlap = intersection.width;
        int yOverlap = intersection.height;

        if (diff.x < 0) { //other is to my right
            xOverlap *= -1;
        }

        if (diff.y < 0) { //other is below me
            yOverlap *= -1;
        }

        boolean canMoveX = (startX + xOverlap > 0 && (startX + xOverlap + width) < thisSection.getWidth());
        boolean canMoveY = (startY + yOverlap > 0 && (startY + yOverlap + height) < thisSection.getHeight());

        int myCentreX = startX + (width / 2);
        int myCentreY = startY + (height / 2);

        int sectionCentreX = thisSection.getWidth() / 2;
        int sectionCentreY = thisSection.getHeight() / 2;

        boolean towardCentreX = ((myCentreX < sectionCentreX && xOverlap > 0) ||
                                    myCentreX > sectionCentreX && xOverlap < 0);

        boolean towardCentreY = ((myCentreY < sectionCentreY && yOverlap > 0) ||
                                myCentreY > sectionCentreY && yOverlap < 0);

        if ((canMoveX) && (canMoveY)) {
            if (Math.abs(xOverlap) > Math.abs(yOverlap)) {
                yOverlap = 0;
            } else {
                xOverlap = 0;
            }
        } else if (canMoveX) {
            yOverlap = 0;
        } else if (canMoveY) {
            xOverlap = 0;
        }

//        xOverlap = diff.x; //Old spreading
//        yOverlap = diff.y;

        if (startX + xOverlap > 0 && (startX + xOverlap + width) < thisSection.getWidth()) {
            if (startY + yOverlap > 0 && (startY + yOverlap + height) < thisSection.getHeight()) {

                Movement thisMovement = new Movement(new Point(startX, startY), new Point(startX + xOverlap, startY + yOverlap));

                if (movementHistory.contains(thisMovement)) {
                    movementHistory.get(movementHistory.indexOf(thisMovement)).increment();
                    thisMovement = movementHistory.get(movementHistory.indexOf(thisMovement));
                } else {
                    movementHistory.add(thisMovement);
                }

                if (thisMovement.getTimesMade() > 5) {
                    Logs.i("Room", "Rejecting " + getId() + " because it's bouncing", null);
                    isRejected = true;
                    return false;
                }

                startX += xOverlap;
                startY += yOverlap;

                timesShifted ++;
                return true;
            }
        }

        return false;
    }

    public Point myCenter() {
        return new Point(startX + (width / 2), startY + (height / 2));
    }

    public Point myGlobalCenter() {
        return new Point(getGlobalStartX() + (width / 2), getGlobalStartY() + (height / 2));
    }

    public double calcHowMuchItFailedBy() {
        double lowerBound = MEAN_ROOM_SIZE - STD_DEV;
        double upperBound = MEAN_ROOM_SIZE + STD_DEV;

        double widthDiff = 0;
        double heightDiff = 0;

        if (width > upperBound) widthDiff = Math.abs(width - upperBound) / upperBound;
        if (width < lowerBound) widthDiff = Math.abs(lowerBound - width) / lowerBound;

        if (height > upperBound) heightDiff = Math.abs(height - upperBound) / upperBound;
        if (height < lowerBound) heightDiff = Math.abs(lowerBound - height) / lowerBound;

        return (widthDiff + heightDiff) / 2;
    }

    public void forceSelect() {
        isRejected = false;
        isSelected = true;
    }

    private boolean willIBeRejected() {
        int lowerBound = MEAN_ROOM_SIZE - STD_DEV;
        int upperBound = MEAN_ROOM_SIZE + STD_DEV;

        if (width < upperBound && width > lowerBound && height < upperBound && height > lowerBound) {
            return false;
        }

        return true;
    }

    public void calculateRejected(List<Room> rooms) {
        for (Room other : rooms) {
            if (this == other) continue;
            if (other.isRejected) continue;

            if (me().intersects(other.me())) {
                if (getArea() > other.getArea()) {
                    isRejected = true;
                } else {
                    other.isRejected = true;
                }
            }
        }

        isSelected = !willIBeRejected();
    }

    public int intersects(Room r) {
        double totalCoverage = Math.min(r.getArea(), getArea());

        Rectangle me = new Rectangle(startX - 3, startY - 3, width + 6, height + 6);
        Rectangle other = new Rectangle(r.startX, r.startY, r.width, r.height);

        Rectangle intersection = me.intersection(other);

        if (intersection.width < 0 || intersection.height < 0) return 0;

        double actualCoverage = me.intersection(other).height * me.intersection(other).width;

        return (int) (100 * (actualCoverage / totalCoverage));
    }

    public List<RoomFeature> getFeatureList() {
        return featureList;
    }

    public void generateFeatures(Modifier modifier) {
        if (willIBeRejected()) return;

        //monster chance
        double monChance = rnd.nextDouble();
        if (monChance <= DEFAULT_MONSTER_CHANCE * modifier.getMonsterChanceModifier()) {
            featureList.add(new MonsterFeature(String.format("%.0f", (monChance * Math.pow(10, 16))), modifier));
        }

        double trapChance = rnd.nextDouble();
        //trap chance
        if (trapChance <= DEFAULT_TRAP_CHANCE * modifier.getTrapChanceModifier()) {
            featureList.add(new TrapFeature(String.format("%.0f", (trapChance * Math.pow(10, 16))), modifier));
        }

        //treasure chance
        double treasureChance = rnd.nextDouble();
        if (rnd.nextDouble() <= DEFAULT_TREASURE_CHANCE * modifier.getTreasureChanceModifier()) {
            featureList.add(new TreasureFeature(String.format("%.0f", (treasureChance * Math.pow(10, 16)))));
        }

        //Stairs Chance
        //As you get further away from the ground floor, the chance halves every floor you are away, to stop infinite growing

        double stairModifier = modifier.stairChanceModifier;
        int level = thisSection.getFloor().getLevel();

        double chanceOfStairFeature = (DEFAULT_STAIRS_CHANCE * stairModifier) / (1 + Math.abs(level));
        double stairChance = rnd.nextDouble();

        if (stairChance <= chanceOfStairFeature) {
            StairsFeature stairsFeature = new StairsFeature(String.format("%.0f", (stairChance * Math.pow(10, 16))), modifier, this);

            featureList.add(stairsFeature);

            stairsFeature.setFloorTarget(thisSection.getFloor().newFloor(stairsFeature.getDirection().getElevation(), this));
        }
    }

    public boolean containsPoint(Point p) {
        return (p.x > startX && p.x < (startX + width) &&
                p.y > startY && p.y < (startY + height));
    }

    public boolean containsGlobalPoint(Point p) {
        return (p.x > getGlobalStartX() && p.x < (getGlobalStartX() + width) &&
                p.y > getGlobalStartY() && p.y < (getGlobalStartY() + height));
    }

    //=========================
    //=========================
    //===== GETTERS ===========
    //=========================
    //=========================
    public int getId() {
        return id;
    }

    public int getGlobalStartX() {
        return startX + sectionStartX;
    }

    public int getGlobalStartY() {
        return startY + sectionStartY;
    }

    public int getStartX() {
        return startX;
    }

    public int getStartY() {
        return startY;
    }

    public int getWidth() {
        return width;
    }

    public int getHeight() {
        return height;
    }

    public int getArea() {
        return width * height;
    }

    public boolean isRejected() {
        return isRejected;
    }

    public boolean isSelected() {
        return isSelected;
    }

    public Rectangle me() {
        return new Rectangle(startX, startY, width, height);
    }

    public Rectangle meWithBorder() {
        return new Rectangle(startX - 2, startY - 2, width + 4, height + 4);
    }

    public Rectangle topBorder() {
        return new Rectangle(startX, startY - 4, width, 6);
    }

    public Rectangle rightBorder() {
        return new Rectangle(startX + width - 2, startY, 6, height);
    }

    public Rectangle leftBorder() {
        return new Rectangle(startX - 4, startY, 6, height);
    }

    public Rectangle bottomBorder() {
        return new Rectangle(startX, startY + height - 2, width, 6);
    }

    public DungeonSection getThisSection() {
        return thisSection;
    }
}

File: app/src/main/java/com/joeshuff/dddungeongenerator/generator/dungeon/Room.kt
package com.joeshuff.dddungeongenerator.generator.dungeon

import com.google.gson.GsonBuilder
import com.joeshuff.dddungeongenerator.db.models.Point
import com.joeshuff.dddungeongenerator.generator.features.*
import com.joeshuff.dddungeongenerator.generator.features.StairsFeature.DIRECTION
import com.joeshuff.dddungeongenerator.generator.floors.DungeonSection
import com.joeshuff.dddungeongenerator.generator.models.Rectangle
import com.joeshuff.dddungeongenerator.db.RuntimeTypeAdapterFactory
import com.joeshuff.dddungeongenerator.util.Logs
import io.realm.RealmObject
import io.realm.annotations.Ignore
import java.util.*
import kotlin.collections.ArrayList
import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.pow
import kotlin.math.sin


open class Room(
        var id: Int,
        @Ignore val thisSection: DungeonSection?,
        @Ignore var rnd: Random?,
        @Ignore var modifier: Modifier?): RealmObject() {

    constructor(): this(0, null, null, null)

    companion object {
        @Ignore
        var featureIntros = Arrays.asList("You've heard rumors that ", "You've discovered that ", "Maps have told you that ", "It became apparent that ", "You know that ")

        @Ignore
        private val MIN_MEAN_ROOM_SIZE = 45

        @Ignore
        private val MAX_MEAN_ROOM_SIZE = 85

        @Ignore
        private val MIN_STD_DEV = 15

        @Ignore
        private val MAX_STD_DEV = 45

        @Ignore
        private var MEAN_ROOM_SIZE = 65

        @Ignore
        private var STD_DEV = 30

        @Ignore
        var DEFAULT_MONSTER_CHANCE = 0.15

        @Ignore
        var DEFAULT_TRAP_CHANCE = 0.2

        @Ignore
        var DEFAULT_TREASURE_CHANCE = 0.25

        @Ignore
        var DEFAULT_STAIRS_CHANCE = 0.05
        fun changeRoomSize(percentage: Float) {
            MEAN_ROOM_SIZE = (MIN_MEAN_ROOM_SIZE + (MAX_MEAN_ROOM_SIZE - MIN_MEAN_ROOM_SIZE) * percentage).toInt()
            STD_DEV = (MIN_STD_DEV + (MAX_STD_DEV - MIN_STD_DEV) * percentage).toInt()
        }
    }

    private var sectionStartX = 0
    private var sectionStartY = 0

    var detail: String = ""

    var startX: Int = 0
    var startY: Int = 0

    @Ignore
    var isSelected = false

    @Ignore
    var isRejected = false

    var width: Int = 0
    var height: Int = 0

    @Ignore
    private var timesShifted = 0

    private var features: String = "{\"features\":[]}"

    fun getUIFeatures(): FeatureContainer {
        val gson = GsonBuilder().registerTypeAdapterFactory(RuntimeTypeAdapterFactory.getAdapter()).create()
        val result = gson.fromJson(features, FeatureContainer::class.java)
        return result
    }

    @Ignore
    private var featureList: ArrayList<RoomFeature> = arrayListOf()

    fun addFeature(roomFeature: RoomFeature) {
        featureList.add(roomFeature)
    }

    @Ignore
    var movementHistory: ArrayList<Movement> = ArrayList()

    init {
        thisSection?.let { section ->
            rnd?.let { random ->
                sectionStartX = section.startPoint?.x?: 0
                sectionStartY = section.startPoint?.y?: 0

                val center = randomPointInCircle(random, section.width / 4)

                height = (random.nextGaussian() * STD_DEV + MEAN_ROOM_SIZE).toInt()
                width = (random.nextGaussian() * STD_DEV + MEAN_ROOM_SIZE).toInt()

                detail = "This room is ${height}ft x ${width}ft\n${RoomDetail.getDetail(random)}"

                startX = (section.width / 2) + (center.x - width / 2)
                startY = (section.height / 2) + (center.y - height / 2)

                modifier?.let { generateFeatures(section, random, it) }
            }
        }
    }

    fun getFullRoomDescription(): String {
        val descParts = getRoomDescription()

        var res = ""

        for (d in descParts) {
            res += "$d\n"
        }

        return res
    }

    fun getRoomDescription(): List<String> {
        val description: ArrayList<String> = ArrayList()

        for (feature in getFeatureList()) {
            description.add(featureIntros[Random().nextInt(featureIntros.size)].toString() + feature.getFeatureDescription())
        }

        return description
    }

    fun connectRoomTo(room: Room, connection: String, direction: DIRECTION) {
        room.thisSection?.floor?.let {
            val newStairs = StairsFeature("", Modifier(), this, direction.opposite(), connection, room.id, it.level)
            addFeature(newStairs)
        }
    }

    private fun randomPointInCircle(random: Random, radius: Int): Point {
        val t = 2.0f * Math.PI * random.nextDouble()
        val u = random.nextDouble() + random.nextDouble()
        val r: Double

        r = if (u > 1) {
            2 - u
        } else {
            u
        }

        val p = Point()
        p.x = (radius * r * cos(t)).toInt()
        p.y = (radius * r * sin(t)).toInt()
        return p
    }

    fun moveAwayFrom(room: Room): Boolean {
        if (thisSection == null) return false

        val center = myCenter()
        val otherCenter = room.myCenter()

        if (timesShifted >= 1000) {
            isRejected = true
            return true
        }

        val diff = Point(center.x - otherCenter.x, center.y - otherCenter.y)

        val intersection = room.meWithBorder().intersection(meWithBorder())

        if (intersection.isEmpty()) return false
        if (diff.x == 0 && diff.y == 0) return false

        var xOverlap = intersection.width
        var yOverlap = intersection.height

        if (diff.x < 0) { //other is to my right
            xOverlap *= -1
        }

        if (diff.y < 0) { //other is below me
            yOverlap *= -1
        }

        val canMoveX = startX + xOverlap > 0 && startX + xOverlap + width < thisSection.width
        val canMoveY = startY + yOverlap > 0 && startY + yOverlap + height < thisSection.height

        val myCentreX = startX + width / 2
        val myCentreY = startY + height / 2

        val sectionCentreX = thisSection.width / 2
        val sectionCentreY = thisSection.height / 2

        val towardCentreX = myCentreX < sectionCentreX && xOverlap > 0 ||
                myCentreX > sectionCentreX && xOverlap < 0

        val towardCentreY = myCentreY < sectionCentreY && yOverlap > 0 ||
                myCentreY > sectionCentreY && yOverlap < 0

        if (canMoveX && canMoveY) {
            if (abs(xOverlap) > abs(yOverlap)) {
                yOverlap = 0
            } else {
                xOverlap = 0
            }
        } else if (canMoveX) {
            yOverlap = 0
        } else if (canMoveY) {
            xOverlap = 0
        }

        if (startX + xOverlap > 0 && startX + xOverlap + width < thisSection.width) {
            if (startY + yOverlap > 0 && startY + yOverlap + height < thisSection.height) {

                var thisMovement = Movement(Point(startX, startY), Point(startX + xOverlap, startY + yOverlap))

                if (movementHistory.contains(thisMovement)) {
                    movementHistory[movementHistory.indexOf(thisMovement)].increment()
                    thisMovement = movementHistory[movementHistory.indexOf(thisMovement)]
                } else {
                    movementHistory.add(thisMovement)
                }

                if (thisMovement.timesMade > 5) {
                    Logs.i("Room", "Rejecting $id because it's bouncing", null)
                    isRejected = true
                    return false
                }

                startX += xOverlap
                startY += yOverlap
                timesShifted++
                return true
            }
        }

        return false
    }

    fun myCenter(): Point {
        return Point(startX + width / 2, startY + height / 2)
    }

    fun myGlobalCenter(): Point {
        return Point(globalStartX + width / 2, globalStartY + height / 2)
    }

    fun calcHowMuchItFailedBy(): Double {
        val lowerBound = MEAN_ROOM_SIZE - STD_DEV.toDouble()
        val upperBound = MEAN_ROOM_SIZE + STD_DEV.toDouble()

        var widthDiff = 0.0
        var heightDiff = 0.0

        if (width > upperBound) widthDiff = Math.abs(width - upperBound) / upperBound
        if (width < lowerBound) widthDiff = Math.abs(lowerBound - width) / lowerBound

        if (height > upperBound) heightDiff = Math.abs(height - upperBound) / upperBound
        if (height < lowerBound) heightDiff = Math.abs(lowerBound - height) / lowerBound

        return (widthDiff + heightDiff) / 2
    }

    fun forceSelect() {
        isRejected = false
        isSelected = true
    }

    private fun willIBeRejected(): Boolean {
        val lowerBound = MEAN_ROOM_SIZE - STD_DEV
        val upperBound = MEAN_ROOM_SIZE + STD_DEV

        return !((width in (lowerBound + 1) until upperBound) && (height in (lowerBound + 1) until upperBound))
    }

    fun calculateRejected(rooms: List<Room>) {
        for (other in rooms) {
            if (this === other) continue
            if (other.isRejected) continue

            if (me().intersects(other.me())) {
                if (area > other.area) {
                    isRejected = true
                } else {
                    other.isRejected = true
                }
            }
        }

        isSelected = !willIBeRejected()
    }

    fun intersects(r: Room): Int {
        val totalCoverage = Math.min(r.area, area).toDouble()

        val me = Rectangle(startX - 3, startY - 3, width + 6, height + 6)
        val other = Rectangle(r.startX, r.startY, r.width, r.height)

        val intersection = me.intersection(other)

        if (intersection.width < 0 || intersection.height < 0) return 0

        val actualCoverage = me.intersection(other).height * me.intersection(other).width.toDouble()

        return (100 * (actualCoverage / totalCoverage)).toInt()
    }

    fun getFeatureList(): List<RoomFeature> {
        return featureList
    }

    fun generateFeatures(section: DungeonSection, random: Random, modifier: Modifier) {
        if (willIBeRejected()) return

        //monster chance
        val monChance = random.nextDouble()
        if (monChance <= DEFAULT_MONSTER_CHANCE * modifier.monsterChanceModifier) {
            addFeature(MonsterFeature(String.format("%.0f", monChance * 10.0.pow(16.0)), modifier))
        }
        val trapChance = random.nextDouble()
        //trap chance
        if (trapChance <= DEFAULT_TRAP_CHANCE * modifier.trapChanceModifier) {
            addFeature(TrapFeature(String.format("%.0f", trapChance * 10.0.pow(16.0)), modifier))
        }

        //treasure chance
        val treasureChance = random.nextDouble()
        if (random.nextDouble() <= DEFAULT_TREASURE_CHANCE * modifier.treasureChanceModifier) {
            addFeature(TreasureFeature(String.format("%.0f", treasureChance * 10.0.pow(16.0))))
        }

        //Stairs Chance
        //As you get further away from the ground floor, the chance halves every floor you are away, to stop infinite growing
        val stairModifier = modifier.stairChanceModifier
        val level = section.floor?.level?: 0

        val chanceOfStairFeature = DEFAULT_STAIRS_CHANCE * stairModifier / (1 + abs(level))
        val stairChance = random.nextDouble()

        if (stairChance <= chanceOfStairFeature) {
            val stairsFeature = StairsFeature(String.format("%.0f", stairChance * 10.0.pow(16.0)), modifier, this)

            section.floor?.newFloor(stairsFeature.direction.elevation, this)?.let {
                addFeature(stairsFeature)
                stairsFeature.setFloorTarget(it)
            }
        }
    }

    fun containsPoint(p: Point): Boolean {
        return p.x > startX && p.x < startX + width && p.y > startY && p.y < startY + height
    }

    fun containsGlobalPoint(p: Point): Boolean {
        return p.x > globalStartX && p.x < globalStartX + width && p.y > globalStartY && p.y < globalStartY + height
    }

    fun complete() {
        val gson = GsonBuilder().registerTypeAdapterFactory(RuntimeTypeAdapterFactory.getAdapter()).create()
        val result = gson.toJson(FeatureContainer(featureList))

        features = result
    }

    //=========================
    //=========================
    //===== GETTERS ===========
    //=========================
    //=========================

    val globalStartX: Int
        get() = startX + sectionStartX

    val globalStartY: Int
        get() = startY + sectionStartY

    val area: Int
        get() = width * height

    fun me(): Rectangle {
        return Rectangle(startX, startY, width, height)
    }

    fun meWithBorder(): Rectangle {
        return Rectangle(startX - 2, startY - 2, width + 4, height + 4)
    }

    fun topBorder(): Rectangle {
        return Rectangle(startX, startY - 4, width, 6)
    }

    fun rightBorder(): Rectangle {
        return Rectangle(startX + width - 2, startY, 6, height)
    }

    fun leftBorder(): Rectangle {
        return Rectangle(startX - 4, startY, 6, height)
    }

    fun bottomBorder(): Rectangle {
        return Rectangle(startX, startY + height - 2, width, 6)
    }
}